### DISCUSSION TITLE
Why alda-player has an expiry ?  And Why Shutdown alda-player when finish play a *.alda file ?

### DISCUSSION THREAD

**Discussion by yueyoum** [alda-lang/alda#497] [Q&A] [answered]
# Q1: Why alda-player has an expiry and will stop its self ?

I run a player process by 

````
$ alda-player -v run -p 9909
uwg INFO  2024-03-23 11:31:31 StateManager.cleanUpStaleStateFiles:88 - Cleaning up stale files in C:\Use
rs\wang\AppData\Local\alda\cache\state\players...
uwg INFO  2024-03-23 11:31:31 StateManager.cleanUpStaleStateFiles:88 - Cleaning up stale files in C:\Use
rs\wang\AppData\Local\alda\cache\state\repl-servers...
uwg DEBUG 2024-03-23 11:31:31 StateManager.cleanUpStaleStateFiles:107 - Deleting empty directory C:\User
s\wang\AppData\Local\alda\cache\state\repl-servers
uwg INFO  2024-03-23 11:31:31 Main.run:77 - Starting receiver, listening on port 9909...
uwg INFO  2024-03-23 11:31:31 MidiEngine.<init>:276 - Initializing MIDI sequencer...
uwg INFO  2024-03-23 11:31:31 MidiEngine.<init>:281 - Initializing MIDI synthesizer...
uwg INFO  2024-03-23 11:31:35 Main.run:90 - Starting player...

````

Then alda ps show the player processes:

```
$ alda ps
id      port    state   expiry  type    pid
uwg     9909    ready   5 minutes from now      player  1936
```

I noticed there has a expiry columu

5 minutes  later

the alda player process will exit its self

```
uwg INFO  2024-03-23 11:37:27 Main.run:104 - Shutting down due to inactivity.
uwg INFO  2024-03-23 11:37:27 Main.invoke:95 - Stopping receiver...
```

----

# Q2: Why shutdown alda-player when finish play a *.alda file ?

```
$ alda play -f test.alda -p 9909
```

the alda-player process show this log and exit its self

```
fqc DEBUG 2024-03-23 11:41:27 MidiEngine._init_$lambda$5:291 - Received CONTINUE meta event
fqc DEBUG 2024-03-23 11:41:27 MidiEngine._init_$lambda$5:321 - Received SHUTDOWN meta event
fqc INFO  2024-03-23 11:41:27 Main.invoke:95 - Stopping receiver...
```

Why this?

I wanna a long life time running alda-player process,  like a background service

I can run many times of `alda play -p 9909`

**Comment by daveyarwood** [comment]
The `alda` client manages player processes in the background, and the player processes ensure that they won't keep running and using up resources when they aren't needed anymore. The idea is that a player process is responsible for playing one score and then shutting itself down.

In normal use of Alda, you should never need to start a player process yourself via `alda-player`. Simply run commands like `alda play -c 'piano: c d e'` and `alda` will spawn player processes as needed to play your scores. The client also sends a "shutdown" message to the player process to schedule a shutdown immediately after the score is finished playing.

**Reply by yueyoum** [reply]
Yes.  I understand and totally agree your design:  

1. keep it simple. (just run alda play or  alda repl ,   it will spawn player process automatically) 
2. gool computer resources management (shutdown when not using)

### BUT  in my usecase:

1.  a mini pc as a dedicated player server.  with a professional HiFi system attached.
2. I use a laptop to run alda client ( alda play -f xxx -p <HiFi PC IP:PORT> )


this usecase has two **problems** right now :

1.  alda only support -p port argument, which means,  the client and server process must running in same computer. 
2. the alda-player, server process will exit its self


Can alda add two feature for support this usecase 

1. add an argument (--host ,  or  --listen,  or --ip )  for specific the server listen address, and client connect target address
2. add an argument for server process running forever

### example:

on server pc:

```
alda-player run --host 0.0.0.0 --port 9999 --forever
```

on client pc

```
alda play -f xxx --host 192.168.1.2 --port 9999
```

with this feature,  I can run a cronjob in linux system,  to make HiFi pc play alda songs

If you agree add this feature,  I will do this and submit a pull request that you can review the code.

**Reply by daveyarwood** [reply, marked as answer]
Player processes are not intended to be long-lived, and you'll run into trouble if you try to use the same player process to play multiple scores. Once a player has played a score, it has that MIDI sequence "loaded" into its internal sequencer, and you could think of it as being in a "dirty" state.

What you're describing is a little involved, but it's actually already possible without needing to make any changes to Alda. Alda's REPL has a client/server architecture, which allows you to run a server on one machine, and send over the code you want to play from another machine.

Ordinarily, when you run `alda repl`, it starts a REPL server in the background, and it presents you with a REPL client session in the foreground. In the client session, you can enter code, as well as other commands, such as `:new` to clear out the REPL server state and initialize a new score. That's all assuming that you want to run everything in one terminal window.

But you can also run an Alda REPL server and client separate from each other, even on different machines. (See `alda repl --help` for options and examples.)

To start just a REPL server, on a specific port, you can run:

```bash
alda repl --server --port 9999
```

You can then use `alda repl --host 1.2.3.4 --port 9999 --message "..."` to send messages to the REPL server. A message is a JSON object with an operation and other data. (There is [comprehensive documentation](https://github.com/alda-lang/alda/blob/master/doc/alda-repl-server-api.adoc) about this API here, if you're curious.)

To play a string of Alda code, you can run a command like this on another machine:

```bash
alda repl \
  --host 192.168.1.2 \
  --port 36521 \
  --message '{"op": "eval-and-play", "code": "piano: c d e"}'
```

There is not an "eval-and-play-file` command (i.e. something analogous to `alda play --file ...`, because of the possibility that the client and server could be running on different hosts. So you would need to send the entire score through as a string in the `code` field, and make sure you properly escape double quotes, etc.  One way to do this is to use the JSON utility [jo](https://github.com/jpmens/jo), which allows you to insert content from a file, and automatically escapes characters like double quotes:

```bash
cat /tmp/quotes.txt 
#=> This is a string with "quotes"

jo normal=string file_contents=@/tmp/quotes.txt
#=> {"normal":"string","file_contents":"This is a string with \"quotes\""}

# Read an Alda source file (this one happens to contain double quotes) and
# insert its contents into the "code" field in the JSON
alda repl \
  --host 192.168.1.2 \
  --port 36521 \
  --message "$(jo op="eval-and-play" code=@/home/dave/code/alda/examples/alternate-endings.alda)"
```

You should be aware that this approach has the same problem as the problem you would have if there were an option to run the same Alda player process "forever" and reuse it to play different scores. The REPL server builds up state over time, so after you use it to play one score, it will interpret any subsequent scores within the context of that original score, which could lead to scores sounding subtly different/wrong.

To solve that problem, you could run the following command to clear out the previous score state and ensure that your next score will be played with a fresh context:

```bash
alda repl \
  --host 192.168.1.2 \
  --port 9999 \
  --message '{"op": "new-score"}'
```

Note that under the hood, the Alda REPL server maintains a small pool of player processes and constantly keeps them alive. When you issue the "new-score" command, it kills the player process that it was using for the previous score and selects a new player from the pool. So it really works almost exactly like `alda play`. But the difference is that the Alda REPL server has an API that is designed for the kind of thing you're doing - networking multiple computers together and controlling playback on one of them.

**Reply by daveyarwood** [reply]
As an alternative solution, you could have your client machine connect via SSH, transfer the score file over, and execute `alda play -f /path/to/the/file.alda` on the server machine. That might actually be a better solution for your simple use case of remotely playing Alda scores.

You could do something like this:

```bash
scp my-score.alda 192.168.1.2:/tmp/my-score.alda
ssh 192.168.1.2 "alda play -f /tmp/my-score.alda"
```

**Reply by yueyoum** [reply]
Thank you for your patience. 


I think you have fixed my trouble.


> Once a player has played a score, it has that MIDI sequence "loaded" into its internal sequencer, 

I'm lack of MIDI technology.  I don't know MIDI player process has this behavior . ( Is this a limitation ? )

So, It looks like a player process ONLY for one *.alda file is a simplest solution .


## BUT I still have other trouble, help me in the new question

**Reply by daveyarwood** [reply]
It's not really a limitation of MIDI. I think if we wanted to, we could throw away the sequence and initialize a new one. But the player processes have other state, and I designed them to be disposed after playing one score, just to avoid unexpected issues with playback.

**Reply by yueyoum** [reply]
By the way,  

If the alda-player process running a thread pool, 

one thread - one midi sequence / midi state

the main thread do the following loop:

1. accept client request
2. get or create a fresh thread
3. dispatch the reqeust to that thread.  and that thread do the actual work.
4. destory that thread when work finished

MIDI State located in thread, NOT entire process

SO ONLY thread need to shutdown,  NOT the process,    

Is this will be a forever running process ?

**Reply by daveyarwood** [reply]
That is essentially exactly what we're doing, except instead of 1 player process with multiple threads, it's one computer with multiple player processes :)

I discovered early on that because the MIDI operations are pretty CPU-heavy, it's better not to have multiple MIDI sequencers/synthesizers running in the same process. Instead, we have this setup where each player process has one MIDI sequencer and synthesizer, and I've found that it works much better.

**Reply by yueyoum** [reply]
Maybe I'm not fully understood the articles you explained to me.

multiple processes do the following loop:

1. main process accept client request
2. fork a child process to do the actual work ( initialize the midi sequencers/synthesizers,  play the alda song)
3. destory that process when work finished

the midi state ONLY bounded in that process,   
main process not do midi things,  ONLY manage child worker (child process)

So, the main process DO NOT MUST quit self,   there will be a forever running player process ?

**Reply by daveyarwood** [reply]
That's correct, except for the last part. Each player process will shut itself down after a period of inactivity.

**Reply by yueyoum** [reply]
So ,,,,

Can we add a argument  that keep the main player process running forever ?

Sorry for my  stubborn,   :)
I just wanna ensure the alda-player process whether can be running forever, 
and can you support this .

**Reply by daveyarwood** [reply]
No, because player processes are not intended to be long-lived, as I explained above.

**Reply by yueyoum** [reply]
I'm totally confused  or maybe I misunderstood.

At the beginning.  you said that MIDI has a state,  So exit the process is a reasonable way.

but then,  the discussion told me, If the main process ONLY manage child player process. 
child player process exit after play one score, 
the main process **DO DOT NEED** exit itself. 

So,  Why not let the main process long-lived ?

Consume A lots of Computer Resource ?

long-lived is not default behavior,  it's a argument,  

If you worry about the computer resource,  run the player process without the loing-live argument.

**Reply by daveyarwood** [reply]
I think you're getting confused about terminology. Let me try to clear it up.

Let's not use the term "main" process. It's not precise enough.

Alda has 3 kinds of processes: the client, the REPL server, and the player.

The client is `alda`. It handles all of the parsing and business logic involved with an Alda score. When you run a command like `alda play -c "piano: c d e"`, it parses the code, creates a score object with a piano part and 3 notes, it spawns player processes if it needs to, it selects a player process, and it sends instructions to the player process to tell it to do the playback.

The player process is `alda-player`. Most Alda users will never need to start it manually or even know that it exists. The client starts player processes automatically to do the playback. The player process starts up, waits for instructions, and follows the instructions. Based on the instructions, it loads notes into a MIDI sequencer, does other MIDI things, and it plays back the MIDI sequence. The player process shuts itself down after playing a score or after a period of inactivity. 

The REPL server provides a way to use Alda more interactively. It's the part of the client that runs when you use the command `alda repl --server`. It receives code like "piano: c d e" from a REPL client, then it uses the same code as `alda play` to parse the code, create a score object, and send instructions to a player process. But the difference is that it doesn't exit after that. It keeps the score object in memory, and when it receives more code as input, like "f g a b > c", it updates the same score object, adding notes to the same piano part.

Does that make sense so far?

**Reply by daveyarwood** [reply]
One of my design goals for this system was that users won't accidentally end up with processes running in the background forever that they aren't aware of. That's why player processes shut themselves down after a period of inactivity.

But the REPL server _will_ run forever until you shut it down explicitly. So if that's the behavior you want, you could explore that approach.

Or, if you want to ensure that an Alda player process is always available, you could set up a cron job or something that runs `alda` every few minutes. Each time you run almost any `alda` command, it automatically checks how many player processes are available and spawns more, if needed. That way, the next time you run an `alda play` command, the playback will be immediate.

**Reply by yueyoum** [reply]
Thank you again for your patient explanation (I even think we should have a face-to-face conversation :) )

I think I roughly understand where our differences lie. 

Your perspective:

1. Users should only need to concern themselves with alda repl or alda play, without having to worry about alda-player. 
2. Moreover, without the users' knowledge, alda-player should not become a persistent service.

My perspective:

1. I completely agree with your design.
2. I want to run a persistent alda-player service on a remote computer, and other computer can connect to that persistent service using alda play --host.
3. The persistent service is a new command-line argument, and only when this argument is provided will alda-player become a persistent service. It will not alter the current default behavior of alda-player.

Just as you said, most users and use cases will not need to interact with alda-player.

However, I hope to provide the option for some professional users, or in special scenarios, to run alda-player separately and have more extensive control over the behavior of alda-player.

**Reply by daveyarwood** [reply]
The good news is that Alda is open source, so if you really want to run an alda-player process persistently, you can clone the repo, comment out [these lines](https://github.com/alda-lang/alda/blob/ce35413618928213daade52a4c9e49393e15e053/player/src/main/kotlin/io/alda/player/Main.kt#L103-L105) and [these lines](https://github.com/alda-lang/alda/blob/ce35413618928213daade52a4c9e49393e15e053/player/src/main/kotlin/io/alda/player/Player.kt#L409-L410), and run `player/bin/build` to build your own player executable with the behavior that you want :)

I'll warn you again, though, that player processes are not intended to play more than one score, and you will most likely encounter unexpected behavior due to accumulated state from previous scores.

**Reply by yueyoum** [reply]
I thought I had already replied, but I just realized now that I can't find my replied content. Maybe I forgot to click the Reply button at that time.


So, there is my reply

----------------------------

I looked at the code link you provided,

I then realized, after our long discussion, that it was a misunderstanding of a noun.

We have been talking about the term "alda-player",

and then we discussed forking multiple alda-player instances.

Although we have been using "alda-player" or "player process" in our discussion,
in fact, we were not talking about the same thing at all.


### Your meaning:

```
- alda-player process 1
- alda-player process 2
- alda-player process 3
```

These player processes are all peers, with equivalent functionality, and can fork themselves.


### My meaning:

```
 alda-manager (main process):
    |- alda-player process 1 (child process)
    |- alda-player process 2 (child process)
    |- alda-player process 3 (child process)
```

What I meant that the running forever process is the **alda-manager process**. 
It doesn't do MiDi things, just accepts requests from clients, forks out alda-players, and the real work is still done by the alda-players. After playing a piece of music, they exit. 
This child process is what you referred to as alda-player. Its behavior hasn't changed.

What I'm saying is to add an alda-manager process, which is the one that runs for a long time. 
the alda-manager process is exactly what I mean.

In this way, this alda-manager will not encounter the problem.


The misunderstanding lies in that you thought I wanted the child processes not to exit. Actually, that's not the case. The child process, as you mentioned, is the alda-player that plays MIDI.


Have I made myself clear now?

**Reply by daveyarwood** [reply]
Yes, that makes sense. I think the missing piece of information here is that `alda` client processes are very good at managing player processes. Any time you run any `alda` command (e.g. `alda play`, or even just `alda` or `alda --help`), the client automatically checks to see how many player processes are running, and it starts more processes as needed. The end result is that there is always a player process available to play your score, right away, without needing to wait for one to start.

I designed it this way deliberately in order to simplify the end user experience. End users don't need to make sure there is some server or manager process running in order to use Alda. They can just run `alda play` commands and it Just Works.
